### PR TITLE
Adapt global gql identification system

### PR DESCRIPTION
**Pull request by hochiw**
- Implemented standard node id interface
- Specify Block and Test object ID as NodeID
- Updated createTest mutation to return whole object

refs #29

### CODE REVIEW

**Comment by chihimng** [comment]
@hochiw please help introduce Node and PageInfo base type, and extend Node type: https://relay.dev/docs/guides/graphql-server-specification/

**Comment by hochiw** [comment]
@chihimng updated 🙏

File: graphql-server/pkg/utils/node_id.go [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+
+	graphql "github.com/99designs/gqlgen/graphql"
+	gqlerrcode "github.com/99designs/gqlgen/graphql/errcode"
+	"github.com/vektah/gqlparser/v2/gqlerror"
+
+	"github.com/oursky/likedao/pkg/models"
+)
+
+// MarshalNodeID is used by gqlgen to write ID type to output.
+func MarshalNodeID(nodeID models.NodeID) graphql.Marshaler {
```
Comment (chihimng) [inline graphql-server/pkg/utils/node_id.go:15]: move to model file?

File: graphql-server/gqlgen.yml [hunk 2]
```diff
@@ -50,11 +50,7 @@ models:
     model: github.com/99designs/gqlgen/graphql.String
     
   ID:
-    model:
-      - github.com/99designs/gqlgen/graphql.ID
-      - github.com/99designs/gqlgen/graphql.Int
-      - github.com/99designs/gqlgen/graphql.Int64
-      - github.com/99designs/gqlgen/graphql.Int32
+    model: github.com/oursky/likedao/pkg/utils.NodeID
```
Comment (chihimng) [inline graphql-server/gqlgen.yml:51]: models.NodeID?

File: graphql-server/pkg/resolvers/test.go [hunk 3]
```diff
@@ -21,17 +21,22 @@ func (r *mutationResolver) CreateTest(ctx context.Context, input models.CreateTe
...
 	}
 
 	return res, nil
 }
 
-func (r *queryResolver) QueryTestsByIds(ctx context.Context, ids []string) ([]*models.Test, error) {
-	res, errs := pkgContext.GetDataLoadersFromCtx(ctx).Test.LoadAll(ids)
+func (r *queryResolver) QueryTestsByIds(ctx context.Context, ids []models.NodeID) ([]*models.Test, error) {
+	var objIds []string
+	for _, id := range ids {
```
Comment (chihimng) [inline graphql-server/pkg/resolvers/test.go:35]: refactor to helper for reuse in the future?

map []NodeID to []string

**Review by chihimng — CHANGES_REQUESTED** [review]

**Comment by hochiw** [comment]
@chihimng updated :adore2x:

**Comment by chihimng** [comment]
@hochiw got extra commit again :sadfrog:

**Comment by hochiw** [comment]
> @hochiw got extra commit again :sadfrog:

removed

File: graphql-server/pkg/models/node_id.go [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+	})
+}
+
+// UnmarshalNodeID is used by gqlgen to read ID from input.
+func UnmarshalNodeID(v interface{}) (NodeID, error) {
+	if tmpStr, ok := v.(string); ok {
+		nid, err := ParseNodeID(tmpStr)
+		if err != nil {
+			gqlerr := gqlerror.Errorf(`invalid node ID "%s"`, tmpStr)
+			gqlerrcode.Set(gqlerr, "VALIDATION_FAILED")
```
Comment (chihimng) [inline graphql-server/pkg/models/node_id.go:33]: can this error code be refactored?

File: graphql-server/pkg/models/node.go [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+	"errors"
+	"fmt"
+	"strings"
+)
+
+var (
+	ErrInvalidNodeID = errors.New("Invalid node ID")
+)
+
+type NodeID struct {
```
Comment (chihimng) [inline graphql-server/pkg/models/node.go:13]: can `node.go` and `node_id.go` be combined? seems both are about NodeID

File: graphql-server/pkg/resolvers/relay.go [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+import (
+	"context"
+	"fmt"
+
+	"github.com/oursky/likedao/pkg/models"
+)
+
+func (r *queryResolver) Node(ctx context.Context, id models.NodeID) (models.Node, error) {
+	switch id.EntityType {
+	case "test":
```
Comment (chihimng) [inline graphql-server/pkg/resolvers/relay.go:15]: how about block?

**Review by chihimng — CHANGES_REQUESTED** [review]

File: graphql-server/pkg/resolvers/relay.go (hunk 6 above)
Comment (hochiw) [reply graphql-server/pkg/resolvers/relay.go:15]: Added a query

**Comment by hochiw** [comment]
@chihimng updated :adore2x:

File: graphql-server/pkg/resolvers/block.go [hunk 7]
```diff
@@ -17,7 +17,14 @@ func (r *queryResolver) GetLatestBlock(ctx context.Context) (*models.Block, erro
 		return nil, err
 	}
 	return res, nil
+}
 
+func (r *queryResolver) QueryBlockByHash(ctx context.Context, hash models.NodeID) (*models.Block, error) {
+	res, err := pkgContext.GetQueriesFromCtx(ctx).Block.QueryBlockByHash(hash.ID)
```
[the remark this answers is not in the document]
Comment (chihimng) [reply graphql-server/pkg/resolvers/block.go:23]: https://github.com/oursky/likedao/issues/52

**Review by chihimng — APPROVED** [review]

**Pull request merged**